homeassistant/helpers/entity.py

```diff
@@ -315,9 +317,52 @@ def unique_id(self) -> str | None:
         """Return a unique ID."""
         return self._attr_unique_id
 
+    @callback
+    def async_full_name(self, device_entry: dr.DeviceEntry) -> str | None:
```

Comment:
I don't see how a full name is a backend thing? It's up to the frontend to decide how to display an entity (eg prefixes or having additional information). This should not be decided or provided by a backend program

Reply:
We need it in backend too to get sane initial entity_ids

Reply:
Agreed, but right now, it is directly used by the `name` property. IMHO, if we need to generate entity IDs, we need to teach the entity ID generation to incorporate/take the device name into account (and not silently add it back into the `name` property).

Reply:
The entity_id generation has indeed been modified in this PR 👍

By having the `name` property return `device_name` + `entity_name` we can ease in this new feature without breaking things everywhere when integrations are updated to implement `entity_name`.

We should also expose the `entity_name` via the entity registry and allow it to be overridden by user, that can be done either in this PR or in a follow-up.

Reply:
> By having the name property return device_name + entity_name we can ease in this new feature without breaking things everywhere when integrations are updated to implement entity_name.

? I don't follow that one. The name property never had a full name, now it has in this PR; I don't see this adding to backward compatibility? Did we add prefix name with device in the core?

Reply:
Let's be careful calling anything "full" and instead be explicit what is included: device + entity name. "full" might indicate the area name is in there too. 

I think that for this PR, we don't need this method at all. The entity platform should keep using the `name` property for entity ID generation, since `name` will return the "device + entity name" if available, or fall back to the existing name.

Reply:
> Let's be careful calling anything "full"

OK, how about `render_composite_name` then?

> The entity platform should keep using the name property for entity ID generation

No, that's not possible because we need the name to influence the `entity_id` generation whereas the `name` property need the entity to registered in the entity registry.

Reply:
> I don't see this adding to backward compatibility

In which way does it not do that?

Reply:
Well, we don't do that now either, so it's not "backwards compatibility".
If anything, it's breaking.

The whole point of the naming changes is not to interfere with naming anymore (let the entity have a name, the device have a name, an area have a name) and have the UI/UX device in the context needed. This still adds/pollutes the actual entity name with knowledge from another domain (device).

Reply:
Had a chat with emontnemery, I completely overlooked the **entity_**name on entity description handling in the full name handling. So that is my bad (if anything, the names are maybe too close to distinguish nicely).

I do think its odd we prefix the device name and not expose it, while that is a thing we are looking to solve in the frontend.